syncnotification.go

```diff
@@ -34,99 +31,382 @@ func (lw *LibWallet) spvSyncNotificationCallbacks(loadedWallet *wallet.Wallet) *
+					if synced {
+						syncProgressListener.OnSyncCompleted()
+					} else {
+						syncProgressListener.OnSyncCanceled()
+					}
 				}
 			})
 		},
-		FetchMissingCFiltersStarted: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchMissingCFilters(0, 0, SyncStateStart)
-			}
-		},
-		FetchMissingCFiltersProgress: func(missingCFitlersStart, missingCFitlersEnd int32) {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchMissingCFilters(missingCFitlersStart, missingCFitlersEnd, SyncStateProgress)
-			}
-		},
-		FetchMissingCFiltersFinished: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchMissingCFilters(0, 0, SyncStateFinish)
-			}
-		},
+		FetchMissingCFiltersStarted:  func() {},
+		FetchMissingCFiltersProgress: func(missingCFitlersStart, missingCFitlersEnd int32) {},
+		FetchMissingCFiltersFinished: func() {},
 		FetchHeadersStarted: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchedHeaders(0, 0, SyncStateStart)
+			if lw.beginFetchTimeStamp != -1 {
+				// already started headers fetching
+				return
 			}
-		},
-		FetchHeadersProgress: func(fetchedHeadersCount int32, lastHeaderTime int64) {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchedHeaders(fetchedHeadersCount, lastHeaderTime, SyncStateProgress)
+
+			lw.beginFetchTimeStamp = time.Now().Unix()
+			lw.startHeaderHeight = lw.GetBestBlock()
+			lw.totalFetchedHeadersCount = 0
+
+			if lw.showLogs && lw.syncing {
+				walletBestBlockTime := lw.GetBestBlockTimeStamp()
+				totalHeadersToFetch := lw.estimateBlockHeadersCountAfter(walletBestBlockTime)
+				log.Infof("Step 1 of 3 - fetching %d block headers.\n", totalHeadersToFetch)
 			}
 		},
+		FetchHeadersProgress: lw.fetchHeadersProgress,
 		FetchHeadersFinished: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnFetchedHeaders(0, 0, SyncStateFinish)
+
+			lw.startHeaderHeight = -1
+			lw.headersFetchTimeSpent = time.Now().Unix() - lw.beginFetchTimeStamp
+
+			// If there is some period of inactivity reported at this stage,
+			// subtract it from the total stage time.
+			lw.headersFetchTimeSpent -= lw.totalInactiveSeconds
+			lw.totalInactiveSeconds = 0
+
+			if lw.headersFetchTimeSpent < 150 {
+				// This ensures that minimum ETA used for stage 2 (address discovery) is 120 seconds (80% of 150 seconds).
+				lw.headersFetchTimeSpent = 150
 			}
-		},
-		DiscoverAddressesStarted: func() {
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnDiscoveredAddresses(SyncStateStart)
+
+			lw.totalFetchedHeadersCount = 0
+
+			if lw.showLogs && lw.syncing {
+				log.Info("Fetch headers completed.")
 			}
 		},
+		DiscoverAddressesStarted: lw.discoverAddressesStarted,
```

Comment:
Might be cleaner to maintain this pattern for **all** callback functions, create a separate `lw.callBackHandler` function to host the method body instead of placing the method body here directly.